[Tan Chen Xiang] iP

xMashedxTomatox opened the pull request:
## DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([Source](https://dansilvestre.com/productivity-quotes/))
DukePro frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- ~~EASY~~ VERY EASY to use
All you need to do is,
1. download it from [here](https://github.com/xMashedxTomatox/ip/releases) 
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you! :+1:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders   

##

If you Java programmer, you can expand on the program too! It is as simple as adding a new command in the `initializeCommand` method, like this!
```
protected void initializeCommands() {
        parser.addCommand(new DeleteCommand());
        parser.addCommand(new ExitCommand());
        parser.addCommand(new FindCommand());
        parser.addCommand(new ListCommand());
        parser.addCommand(new MarkCommand(true));
        parser.addCommand(new MarkCommand(false));
        parser.addCommand(new StoreDeadlineCommand());
        parser.addCommand(new StoreEventCommand());
        parser.addCommand(new StoreTodoCommand());
}
```

Commit:
Add Gradle support

xMashedxTomatox committed:
Add function greet, echo and bye

xMashedxTomatox committed:
Add function store, list

xMashedxTomatox committed:
Update function list

xMashedxTomatox committed:
Add functionalities to mark and unmark tasks.

Added class Task. Added interface ICompletable. Added functions mark, unmark. Updated function list.

xMashedxTomatox committed:
Updated class hierarchy and file structure.

xMashedxTomatox committed:
Add inheritance hierarchy and output

Added sub classes Deadline, Event and Task.
Updated output of function list.

xMashedxTomatox committed:
Add test case and bug fixes

Added a test case and used it to identify a bug.
Bugfixes
- Removed an additional echo

xMashedxTomatox committed:
Add exceptions and error handling, Update test cases

Added exceptions and invalid input handling for most of the foreseeable errors. Also added additional  invalid input in test case.

xMashedxTomatox committed:
Add function delete. Update test cases.

Added functionalities for deleting tasks, also updated test case to include deletion of task and additional invalid input.

xMashedxTomatox committed:
Add Storage, Update framework

Added a storage class that handles the serialization/deserialization of tasks. Updated framework to use existing task list.

xMashedxTomatox committed:
Merge branch 'branch-Level-7'

xMashedxTomatox committed:
Add date time implementation

Modified the implementation of the deadline and event classes to use date time instead of strings.

xMashedxTomatox committed:
Update saving occurrence

Updated the system to save whenever the task list changes rather than only saving at the end of the program

xMashedxTomatox committed:
Merge branch 'branch-Level-8'

xMashedxTomatox committed:
Update framework to be more OOP

Updated the framework by extracting out the functionalities of the DukeManager class into other classes. Added UI, Parser, Command and its subclasses.

xMashedxTomatox committed:
Merge branch 'branch-A-MoreOOP'

xMashedxTomatox committed:
Add packages and class relocation

Added new packages Commands, Exception, External, Managers, Tasks. Separated all existing classes into respective packages. 
Bug Fixes
- Added the exit command

xMashedxTomatox committed:
Merge branch 'branch-A-Packages'

xMashedxTomatox committed:
Rename all packages to lower case

Renaming all packages to start with a lower case

xMashedxTomatox committed:
Update OOP structure

Updated OOP structure by extracting the todo, deadline and event commands into their respective subclasses.

xMashedxTomatox committed:
Add JUnits for Parser and TaskList class

Added JUnit for Parser class, with 2 function testing for valid/invalid input for the parse functions.
Added JUnit for TaskList class, testing for valid/invalid input for the respective tasks creation.

xMashedxTomatox committed:
Merge branch 'branch-A-JUnit'

xMashedxTomatox committed:
Update output of error messages

Updated output of error messages to end with a line separator.

xMashedxTomatox committed:
Add MF file for jar creation

added MANIFEST.MF for the creation of v0.1 release

xMashedxTomatox committed:
Add javadocs for functions of source classes

Added documentation for all source classes and their public functions.

xMashedxTomatox committed:
Update program to follow coding standard

Made changes to names in the Command and ExitCommand classes to follow the given coding standard.

xMashedxTomatox committed:
Add find command

Updated functionality of parser and task list to incorporate a new find command
Added FindCommand class.

xMashedxTomatox committed:
Update task description name

Updated the variable name that represents the description of a tasks from info to description.

xMashedxTomatox committed:
Merge branch 'branch-A-JavaDoc'

xMashedxTomatox committed:
Merge branch 'branch-A-CodingStandard'

xMashedxTomatox committed:
Merge branch 'branch-Level-9'

xMashedxTomatox committed:
Add new JUnit test

Added new JUnit tests for the new find commands

xMashedxTomatox committed:
Update abstraction of TaskList

Updated the abstraction of the TaskList class by changing all the values returned by the functions of TaskList from Task to String.

xMashedxTomatox committed:
Add javadocs for classes in duke.tasks package

Added javadocs for all the classes in the duke.tasks package.

xMashedxTomatox committed:
Refactor code structure for ease of future extensions.

Refactored the structure of the system to be more modular so as to make the program scale easier.

joshuayapwj98 commented on src/test/java/duke/managers/ParserTest.java:
```diff
@@ -0,0 +1,98 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ParserTest {
```
Love the extensive test cases 😄

joshuayapwj98 commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,47 @@
+        return "[T]" + super.toString();
+    }
+
+    /**
+     * The factory function of the todo task. Throws an exception
+     * if the provided tokens does not satisfy the required format to create
+     * a todo task.
+     *
+     * @param tokens the specification of the todo task which contains
+     *               the description of the task.
+     */
+    protected static Todo createTask(String[] tokens) throws DukeException {
+        String item = "";
```
The summary of the method is concise!

However, a @return is missing here!

Small issue 😄

joshuayapwj98 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,135 @@
+import java.io.Serializable;
+import java.util.ArrayList;
+
+import duke.tasks.*;
```
Just curious if all of the tasks are required to be imported!

anyway, very neat stuff!! Good job! 👍🏼

joshuayapwj98 reviewed (commented):
Overall, a very neat and easy to read patch of code.

Love the conciseness and accuracy of code, taking into detail all the comments that you have for all the methods in each class. Despite minor corrections, almost a perfect implementation of the given iP task!  Strive on with the awesome work! 💯

yongler commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,49 @@
+import duke.exceptions.DukeException;
+import duke.tasks.TaskList;
+import duke.managers.Ui;
+import duke.managers.Storage;
```
Maybe some spacing here would increase readability too!

yongler commented on src/main/java/duke/managers/DukeManager.java:
```diff
@@ -0,0 +1,78 @@
+package duke.managers;
+
+import duke.commands.*;
```
This works too, but Imports in explicit form is recommended

yongler commented on src/main/java/duke/managers/Parser.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class Parser {
+
+    protected ArrayList<Command> commandList;
```
Maybe Plural form should be used on names representing a collection of objects? e.g. commands

yongler commented on src/test/java/duke/managers/ParserTest.java:
```diff
@@ -0,0 +1,98 @@
+            //test case 8
+            assertTrue(parser.parse("unmark 1") instanceof MarkCommand);
+
+            //test case 9
```
Really extensive test cases!

yongler reviewed (commented):
Looks good overall, LGTM!

Dineshraj555 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public String toString() {
+        if (hasTime) {
+            return "[D]" + super.toString() + " (by: " + date.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + " " + time.format(DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)) + ")";
```
Maybe consider breaking line 48 into 2 separate lines as there are more than 120 characters.

Dineshraj555 reviewed (commented):
LGTM. Well written. One minor suggestion on top of what has already been mentioned by the others.

xMashedxTomatox committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

xMashedxTomatox committed:
Add gradle support for building and JUnit

Added gradle support for building of jar files and JUnit tests.

xMashedxTomatox committed:
Merge branch 'branch-A-Gradle'

xMashedxTomatox committed:
Add checkstyle and suppression rules

Added xml files providing the rules for checkstyle and suppression.

xMashedxTomatox committed:
Fix all checkstyle errors and warnings

Updated code base to be inline with the provided checkstyle. 
Removed external io script.

xMashedxTomatox committed:
Merge branch 'branch-A-CheckStyle'

xMashedxTomatox committed:
Add GUI functionalities

Added GUI functionalities to the program and updated all classes to support it.

xMashedxTomatox committed:
Merge branch 'branch-Level-10'

ianfromdover commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,31 @@
+package duke;
+
+import java.io.IOException;
+
+import duke.internal.Duke;
+import duke.ui.MainWindow;
+import javafx.application.Application;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
```
Perhaps you may want to move the duke imports right under the package instead :) Just a minor suggestion

ianfromdover commented on src/main/java/duke/commands/HelpCommand.java:
```diff
@@ -0,0 +1,61 @@
+            + "7. mark <valid task number> (marks the task with the given task number as done)" + "\n"
+            + "8. mark <valid task number> (unmarks the task with the given task number as undone)" + "\n"
+            + "9. find <search info> (displays all tasks with info similar to the provided details)" + "\n"
+            + "10. bye (exits the program)";
```
Good placement of the `+` operator aids in comprehension 👍

ianfromdover commented on src/main/java/duke/internal/Duke.java:
```diff
@@ -0,0 +1,91 @@
+                e.printStackTrace();
+            }
+            return ui.getFormattedMessage(output);
+        }
```
I found it hard to understand what the nested try-catch block does in this place, is catching the DukeException necessary ? :)

ianfromdover commented on src/main/java/duke/internal/Parser.java:
```diff
@@ -0,0 +1,56 @@
+        String[] tokens = input.split(" ");
+        Command recognizedCommand = null;
+        for (Command c : commands) {
+            if (c.checkIdentifier(tokens[0])) {
```
Perhaps you may like to assign the `tokens[0]` to a variable which gives it meaning to increase readability :)

ianfromdover reviewed (commented):
Overall well organised, adheres closely to coding standards, good job 👍

xMashedxTomatox committed:
Update conventions for factory methods of Deadline, Todo and Event

Factory methods of Deadline, Event and Todo contains different extract-able logic and is very long.

The length of the code exceeds the recommended 30 lines and makes it harder to read.

Extracting the logic for recognizing the different attributes of the respective task from the user input into separate functions. 

Encapsulating the logic for recognizing separate attributes of a task makes the code more concise and readable.

xMashedxTomatox committed:
Update conventions for Duke

Function initialize in Duke class have unclear paths between the erroneous/unusual and happy clauses of the function.

This violates the happy path prominent technique of java conventions and makes the code slightly unreadable as it can distract readers from the alternative paths.

Separated the logic for rare/error clauses (e.g. when a user uses the program initially) from the happy clause.

This makes the logic for the code more defined as it is properly split into separate sections. As a result the code is easier to read and understand.

xMashedxTomatox committed:
Update conventions for Storage

Catch clause of function save does not accurately document it's intended behavior as it is common for this exception to be caught (e.g. when a user starts up for the first time).

Printing stack trace can be misinterpreted as a bug occurring in the program.

Removing the stack trace and leaving a documentation indicating that there is no problem would remove any ambiguity.

Because it is a commonly caught exception that does not raise any problem in the system, leaving a javadoc indicating so is sufficient and clear,

xMashedxTomatox committed:
Add assertions for initialization checks

Added assertions for execution method in certain Task sub classes and Duke.

Task sub classes: Most of the execute functions in tasks requires and assumes that the task list provided is properly initialized. 

Duke functions: Functions in duke requires the internal managers of the duke to be initialized.

These assertions ensures that the functions are only called in the program loop.

xMashedxTomatox committed:
Merge pull request #1 from xMashedxTomatox/branch-A-Assertions

Add assertions for initialization checks

xMashedxTomatox committed:
Merge pull request #2 from xMashedxTomatox/branch-A-CodeQuality

Branch a code quality

xMashedxTomatox committed:
Add sort functionalities for task list

Requirement to be able to sort the task list according to alphabetical and chronological order.

Functionality allows user to sort and view their task in order.

Add two different sorting functions, one for chronological and one for alphabetical.

This is the two most common ways users would want to view their tasks by.

xMashedxTomatox committed:
Merge pull request #3 from xMashedxTomatox/branch-C-Sort

Add sort functionalities for task list

xMashedxTomatox committed:
Update GUI display

Current GUI display looks aesthetically unpleasing, the colors looks awkward and there is no differentiation between error response and normal response.

Unfriendly for user to use as the display makes it hard for user to read the response.

Updating the GUI by adding different color schemes to different types of messages. (e.g. error or normal response). Also changed the padding and margin values to make the display look more refined.

xMashedxTomatox committed:
Update command response

Current command responses is not intuitive enough that the user will be able to figure out what the proper format of the commands are.

User might not be able to use the program as intended due to the lack of information in the response message.

Updating the messages to be clearer by displaying what information is missing/erroneous in the command. Also added examples to make it clearer.

xMashedxTomatox committed:
Merge branch 'branch-A-BetterGui'

xMashedxTomatox committed:
Add User Guide and representative screenshot

Current documentation is insufficient for user to get enough information to use the application effectively.

As such user may be unable to use the application as smoothly as intended.

Adding a user guide and a representative screenshot will give the user something to refer to to get a brief overview and guide of the program.

xMashedxTomatox committed:
Merge branch 'branch-A-UserGuide'

xMashedxTomatox committed:
Set theme jekyll-theme-slate

xMashedxTomatox committed:
Fix bugs in user guide

Current formatting of user guide displays incorrectly in HTML file.

The user guide is not reflecting as intended and can present incorrect information to the user.

Updating the user guide to try to make it display correctly in HTML.

xMashedxTomatox committed:
Set theme jekyll-theme-hacker

xMashedxTomatox committed:
Update storage path

Currently the storage path defined might be ambiguous and cause bugs in certain OSes without properly defined JVM path.

Can cause data to be saved in unintended locations and as a result lose the tasks that has been saved.

Properly defined the saving directory to be in the same directory as the jar file.

xMashedxTomatox committed:
Merge branch 'branch-A-Release'

xMashedxTomatox committed:
Update Storage file saving

Current file saving does not work as intended with directories that contains spaces.

If user saves the jar file in a directory containing spaces, the data file may be saved incorrectly to another directory.

Updating the storage file path to take into account spaces, this should make saving data work as intended regardless of space or not.
